pkg/alertmanager/multitenant.go

```diff
@@ -263,45 +263,47 @@ func (am *MultitenantAlertmanager) loadAllConfigs() map[string]configs.View {
 	}
 }
 
-func (am *MultitenantAlertmanager) updateConfigs(now time.Time) error {
+func (am *MultitenantAlertmanager) updateConfigs() error {
 	cfgs, err := am.poll()
 	if err != nil {
 		return err
 	}
-	am.addNewConfigs(cfgs)
+	am.syncConfigs(cfgs)
 	return nil
 }
 
-// poll the configuration server. Not re-entrant.
-func (am *MultitenantAlertmanager) poll() (map[string]configs.View, error) {
-	configID := am.latestConfig
-	cfgs, err := am.configsAPI.GetAlerts(context.Background(), configID)
+// poll the alert store. Not re-entrant.
+func (am *MultitenantAlertmanager) poll() (map[string]alerts.AlertConfigDesc, error) {
+	cfgs, err := am.store.ListAlertConfigs(context.Background())
 	if err != nil {
-		level.Warn(util.Logger).Log("msg", "MultitenantAlertmanager: configs server poll failed", "err", err)
+		level.Warn(util.Logger).Log("msg", "MultitenantAlertmanager: configs alert store poll failed", "err", err)
 		return nil, err
 	}
-	am.latestMutex.Lock()
-	am.latestConfig = cfgs.GetLatestConfigID()
-	am.latestMutex.Unlock()
-	return cfgs.Configs, nil
+	return cfgs, nil
 }
 
-func (am *MultitenantAlertmanager) addNewConfigs(cfgs map[string]configs.View) {
+func (am *MultitenantAlertmanager) syncConfigs(cfgs map[string]alerts.AlertConfigDesc) {
 	// TODO: instrument how many configs we have, both valid & invalid.
 	level.Debug(util.Logger).Log("msg", "adding configurations", "num_configs", len(cfgs))
-	for userID, config := range cfgs {
-		if config.IsDeleted() {
-			am.deleteUser(userID)
-			continue
-		}
-		err := am.setConfig(userID, config.Config)
+	for _, cfg := range cfgs {
+		err := am.setConfig(cfg)
 		if err != nil {
 			level.Warn(util.Logger).Log("msg", "MultitenantAlertmanager: error applying config", "err", err)
 			continue
 		}
+	}
 
+	am.alertmanagersMtx.Lock()
+	defer am.alertmanagersMtx.Unlock()
+	for user, userAM := range am.alertmanagers {
+		if _, exists := cfgs[user]; !exists {
+			userAM.Stop()
+			delete(am.alertmanagers, user)
+			delete(am.cfgs, user)
+			level.Info(util.Logger).Log("msg", "deleting alertmanager", "user", user)
```

Comment:
I know I'm annoying. Technically this should go before `userAM.Stop()`. Why? Because if any critical issue occurs (ie. a panic) we see what it was doing, otherwise it's not guaranteed the execution reaches this point.

Here may eventually have a `deleted alertmanager` log (given deletions should not be that frequest, looks good knowing when it started and completed).

Reply:
I moved this log line and added a `deleted` log line to clarify